### ISSUE TITLE
Asynchronous programming codelab: consider updates to example code

### ISSUE THREAD

**Issue by chalin** [dart-lang/site-www#2128]
Moving the discussion from https://github.com/dart-lang/site-www/pull/2114#discussion_r348243358 to this issue so that PR #2114 can contain only the original code. Copying the relevant parts of the discussion below.

Also consider that the **analyzer** (with the package using the `pedantic` analyzer options), **raises issues** with quite a few samples of the code, including: `await_only_futures`, `argument_type_not_assignable`, `invalid_assignment`, `type_annotate_public_apis` (which is the most common, but probably the least problematic). This should probably be addressed. In #2114, we used `ignore` directives to silence the analyzer.

---

Here's part of the old code that is relevant to this discussion:

```
void createOrderMessage () async {
  print('Awaiting user order...');
  var order = await getUserOrder();
  print('Your order is: $order');
}

Future<String> getUserOrder() {
  // Imagine that this function is more complex and slow.
  return Future.delayed(Duration(seconds: 4), () => 'Large Latte');
}

main() async {
  countSeconds(4);
  await createOrderMessage();
}
```

Here's the reasoning behind the changes I made, along with comments about what I think is peculiar about the original code:

 1. At the start of (original) [Working with futures: async and await](https://dart.dev/codelabs/async-await#working-with-futures-async-and-await) we write:

    > If [an async] ... function doesn’t explicitly return a value, then the return type is `Future<void>`:
    >
    > ```dart
    > Future<void> main() async {
    > ```

    So, ...
    1. **Shouldn't we be following our own guidelines?**
    2. Why do we given an example of attributing a return type of `Future<void>` to `main()` if we never do so in the codelab?

 2. Pedagogically, I find it strange that as we work through the codelab, `String createOrderMessage()` starts off with a **semantically meaningful name** and consistent return type. Then later we repurpose the function to merely _print the order message_ (rather than return it) and so change the signature to `void createOrderMessage()`. By doing so we're introducing an incongruity. One solution would be to change the function name to `printOrderMessage()`.

3. Regardless of our choice of function name in the previous point (say, `void printOrderMessage()`), by using a return type of `void` we're again not following our own guideline (the guideline I mention in point 1), and we're not explaining why we're deviating from our own guideline.

Thoughts?

---

Btw, when `createOrderMessage()` is declared as `void`, we get the following issue reported by the analyzer:

```nocode
'await' applied to 'void', which is not a 'Future'. dart(await_only_futures)
```

**Comment by legalcodes** [comment]
Thanks for raising this issue, happy to consider changes to this codelab!

Re: `If [an async] ... function doesn’t explicitly return a value, then the return type is Future<void>`, is this the relevant part of the codelab you're referencing? 

![Screen Shot 2019-11-21 at 10 49 55 AM](https://user-images.githubusercontent.com/5550175/69367347-f672fc00-0c4c-11ea-9eb8-4edfd0104dc5.png)

**Comment by chalin** [comment]
Yes

**Comment by legalcodes** [comment]
@galeyang has kindly agreed to review notes about any UX considerations that were relevant here. Outside of that, I'm totally open to adding the type annotations you've suggested. 

Does this capture the proposed changes? 

![futures_intro](https://user-images.githubusercontent.com/5550175/69383765-67c3a680-0c6f-11ea-9a09-30d367ee1ab6.png)
![get order sync bad](https://user-images.githubusercontent.com/5550175/69383766-698d6a00-0c6f-11ea-8a52-2caf25402660.png)
![get_order](https://user-images.githubusercontent.com/5550175/69383771-6abe9700-0c6f-11ea-9f47-6292436c44e4.png)

**Comment by chalin** [comment]
Thanks @galeyang! Thanks for the code diffs Jon, although I'm not sure how clear they are with the docregion tags. PR #2114 now contains the original code, so the **simplest way to visualize the suggested code changes would be to submit a PR with the changes made** (that way we could use the GitHub diff).

That being said, the main suggested code changes are:
 1. Update the code so that it has no analyzer warnings when configured to use the pedantic analyzer options.
    In particular, this would mean (but not be limited to):
    - Always declaring `main()` with a return type.
    - Always declaring async functions returning nothing as `Future<void>` (not `void`, as is done in a few cases).
 2. Don't repurpose `createOrderMessage()`, instead rename the function to, say, `printOrderMessage()`. See (2) in the original comment for details.

Other points are raised in the original comment of this issue.

Also, a minor suggested change is to write `getUserOrder()` as an arrow function.

**Comment by legalcodes** [comment]
Agreed that ideally we can review these changes in a GitHub PR, but since the code still lives in gists, we can't compare updates in the PR to the prior code (i.e. the diff will simply be new code, as opposed to a diff of changes to the old code), or have I missed something @chalin ?

**Comment by legalcodes** [comment]
I'm assuming you meant that *once #2114 is merged* we'll be able to raise the changes in a GitHub PR, which works for me. Just worth noting that we don't have an ETA on merging that PR yet, correct?

**Comment by chalin** [comment]
> Agreed that ideally we can review these changes in a GitHub PR, but since the code still lives in gists, ...

Oh, I'm sorry if it wasn't clear. I just added https://github.com/dart-lang/site-www/pull/2114#issuecomment-557635064 to clarify that #2114 now uses the same code as the Gists, modulo running dartfmt. So the code in the code blocks of `async-await.md` corresponds to the Gist code. (FYI, this is the main commit that reverts code changes: https://github.com/dart-lang/site-www/pull/2114/commits/ed3930ed00bdc50d785caf3f3395e1f137a1f08a.)

> I'm assuming you meant that once #2114 is merged ...

AFAIK, it is possible to submit a PR over a PR (and then to view the new PR's delta, just look at the new PR's single commit diff).

**Comment by galeyang** [comment]
Re: @chalin's initial [comment](https://github.com/dart-lang/site-www/issues/2128#issue-526758837)

`#1:`I'm trying to remember our discussion around the return type declaration before `main()` and found previous [comments](https://github.com/dart-lang/site-www/pull/1659#discussion_r310330259) between @legalcodes, @kwalrath and me. Basically  I found it confusing if we only add `Future<void>` before `main()` in some examples but not _all_ examples, especially when we didn't explain why in that version. Kathy and Jon talked about ` It seems sort of odd to show examples with Future<void> main() everywhere, but it also seems odd to do it in some places but not others.` Eventually we added a blue note to explain that return type declaration being OK to leave off. 

![image](https://user-images.githubusercontent.com/5102864/69450575-52538880-0d12-11ea-8c57-5d21ccca9eba.png)
 
This note was initially put at the end, a user (via survey) suggested us to move it upfront ([issue](https://github.com/dart-lang/site-www/issues/1871#issue-490062863)). I'm fine with either having `Future<void>` consistently or leave it as it is now and ensure we explain why sometimes they are different.

`#2`: I'm fine with this!
`#3`: Good point. I actually don't know why it is not `Future<void> but `void`.

**Comment by galeyang** [comment]
I just found in the [Effective Dart](https://dart.dev/guides/language/effective-dart/design#do-use-futurevoid-as-the-return-type-of-asynchronous-members-that-do-not-produce-values), it has recommendations 

![image](https://user-images.githubusercontent.com/5102864/69452235-0a366500-0d16-11ea-9d4b-c3b1c7b7fc48.png)

**Comment by kwalrath** [comment]
I think that recommendation was pointed at fully typed older code that used either `Future` or `Future<Null>`. Taken literally, it doesn't seem to apply to `main` because `main` isn't a member (afaik).

That said, I think the pedantic package might now complain if we don't type `main`. (Not sure about that. @chalin, do you happen to know?)

**Comment by chalin** [comment]
> the pedantic package might now complain if we don't type `main`.

Yes:
> <img src="https://user-images.githubusercontent.com/4140793/69454007-36141380-0d33-11ea-9c75-1636b7d150ab.png" width=500>

Which is why the Async codelab code in #2114 is sprinkled with `ignore` directives :).

**Comment by legalcodes** [comment]
Summing up, referencing @chalin  's original proposal:

It sounds like we're all OK with:
(1)  declaring `main()` with a return type for all example code in this codelab
(2) Declaring async functions returning nothing as `Future<void>`
(3) Rename `createOrderMessage()` to `printOrderMessage()`. 

So these changes should go in!
Does anyone want to weigh in on the refactor using arrow functions? @kwalrath ?

**Comment by chalin** [comment]
Do we generally agree that codelab, tutorial and pretty much **all example code** should be **free of analyzer issues** when configured with the **pedantic** options?

**Comment by chalin** [comment]
(Unless, of course, it is meant to, e.g.: illustrate errors, or be an incomplete starting point for an exercise.)

**Comment by legalcodes** [comment]
Speaking for myself: yes, absolutely. Thanks Patrice!

**Comment by natebosch** [comment]
+1 for using `Future<void>` as a return type in any place where we will be using an `await` with it.

`void` as a return type is OK only if callers should never await it, which I don't think comes up in this codelab.

+1 for changing the name between `create` and `print` depending on the behavior, return vs print.

Also, see https://github.com/dart-lang/site-www/pull/2134

We should rename all `get*` methods, and ideally make some of them getters.

**Comment by chalin** [comment]
There are still some points from this issue that haven't been addressed. Reopening.

**Comment by legalcodes** [comment]
@chalin Can you specify which points from this issue haven't been addressed?

**Comment by chalin** [comment]
> Can you specify which points ...

- https://github.com/dart-lang/site-www/pull/2137/files#r350924276, which you know about already.
- https://github.com/dart-lang/site-www/pull/2137/files#r351422017

(There might be one more point, will post it here if I find it.)
